src/openzeppelin/account.cairo

```diff
@@ -0,0 +1,218 @@
+    use super::ArrayCallSerde;
+    use super::ArrayCallDrop;
+    use super::ACCOUNT_ID;
+
+    use openzeppelin::introspection::erc165::ERC165Contract;
+
+    //
+    // Storage and Constructor
+    //
+
+    struct Storage {
+        public_key: felt252,
+    }
+
+    #[constructor]
+    fn constructor(_public_key: felt252) {
+        ERC165Contract::register_interface(ACCOUNT_ID);
+        public_key::write(_public_key);
+    }
+
+    //
+    // Externals
+    //
+
+    #[external]
+    fn __execute__(mut calls: Array<Call>) -> Array<Array<felt252>> {
+        assert_valid_transaction();
+        let mut res = ArrayTrait::new();
+        _execute_calls(calls, res)
+    }
+
+    #[external]
+    fn __validate__(mut calls: Array<Call>) {
+        assert_valid_transaction()
+    }
+
+    #[external]
+    fn __validate_declare__(class_hash: felt252) {
+        assert_valid_transaction()
+    }
+
+    #[external]
+    fn __validate_deploy__(
+        class_hash: felt252,
+        contract_address_salt: felt252,
+        _public_key: felt252
+    ) {
+        assert_valid_transaction()
+    }
+
+    #[external]
+    fn set_public_key(new_public_key: felt252) {
+        assert_only_self();
+        public_key::write(new_public_key);
+    }
+
+    //
+    // View
+    //
+
+    #[view]
+    fn get_public_key() -> felt252 {
+        public_key::read()
+    }
+
+    #[view]
+    fn is_valid_signature(message: felt252, sig_r: felt252, sig_s: felt252) -> bool {
+        let _public_key: felt252 = public_key::read();
+        check_ecdsa_signature(message, _public_key, sig_r, sig_s)
+        // to do:
+        // return magic value or false
+    }
+
+    #[view]
+    fn supports_interface(interface_id: felt252) -> bool {
+        ERC165Contract::supports_interface(interface_id)
+    }
+
+    //
+    // Internals
+    //
+
+    fn _execute_calls(mut calls: Array<Call>, mut res: Array<Array<felt252>>) -> Array<Array<felt252>> {
+        match calls.pop_front() {
+            Option::Some(call) => {
+                let _res = _execute_single_call(call);
+                res.append(_res);
+                return _execute_calls(calls, res);
+            },
+            Option::None(_) => {
+                return res;
+            },
+        }
+    }
+
+    fn _execute_single_call(mut call: Call) -> Array<felt252> {
+        let Call{to, selector, calldata } = call;
+        starknet::call_contract_syscall(to, selector, calldata).unwrap_syscall()
+    }
+
+    fn assert_only_self() {
+        let caller = get_caller_address();
+        let self = get_contract_address();
+        assert(self == caller, 'Account: unauthorized.');
+    }
+
+    fn assert_valid_transaction() {
+        let tx_info = unbox(get_tx_info());
+        let tx_hash = tx_info.transaction_hash;
+        let signature = tx_info.signature;
+
+        assert(signature.len() == 2_u32, 'bad signature length');
+
+        let is_valid = is_valid_signature(
+            tx_hash,
```

Comment:
This hash is produced and provided by the protocol. There's [a few functions to compute the right hash](https://github.com/OpenZeppelin/nile/blob/main/src/nile/core/types/utils.py#L47-L99) for each type of transaction.

Specifically for `__execute__`, there's a [de facto standard](https://docs.openzeppelin.com/contracts-cairo/0.6.1/accounts#call_and_accountcallarray_format) followed by all Accounts (OpenZeppelin, Argent, Braavos mainly) based on `Call` and `CallArray`. This will change now with cairo1 where we can simply use `Array<Call>`.